#' Empirical codon usage indexes.
#'
#' This function computes empirical codon usage indexes based on either ribosome
#' P-sites, A-site or E-site frequency associated to in-frame P-sites along
#' coding sequences. Given one sample, it computes 64 codon usage indexes (one
#' for each triplet, optionally normalized for their frequency in CDSs) and
#' generates a bar plot of the resulting values. If two samples are specified,
#' the function also compares the two sets of codon usage indexes returning a
#' scatter plot. The same output is generated specifying one sample and
#' providing 64 triplet-specific values. Scatter plots report the result of
#' linear regressions between the two sets of values and the corresponding
#' Pearson correlation coefficient.
#' 
#' @param data List of data tables from \code{\link{psite_info}}. Each data
#'   table may or may not include one or more columns among \emph{p_site_codon},
#'   \emph{a_site_codon} and \emph{e_site_codon} reporting the three nucleotides
#'   covered by the P-site, A-site and E-site, respectively. These columns can
#'   be previously generated by the \code{\link{psite_info}} function.
#'   Otherwise, the column of interest can be specified by \code{site} and it is
#'   automatically generated starting from a FASTA file or a BSgenome data
#'   package.
#' @param annotation Data table as generated by \code{\link{create_annotation}}.
#' @param sample Either character string or character string vector specifying
#'   one or two sample names, respectively. If one sample name is specified, a
#'   bar plot displaying the 64 codon usage indexes is generated. If two sample
#'   names are specified, the function also compares the two sets of codon usage
#'   indexes returning a scatter plot. In the latter case it also performes a
#'   linear regression and computes the Pearson correlation coefficient.
#' @param site Either "psite, "asite", "esite". It specifies if the empirical
#'   codon usage indexes should be based on ribosome P-sites ("psite"), A-sites
#'   ("asite") or E-sites ("esite"). Default is "psite".
#' @param fastapath Character string specifying the FASTA file used in the
#'   alignment step, including its path, name and extension. This file can
#'   contain reference nucleotide sequences either of a genome assembly or of
#'   all the transcripts (see \code{Details} and \code{fasta_genome}). Please
#'   make sure the sequences derive from the same release of the annotation file
#'   used in the \code{\link{create_annotation}} function. Note: either
#'   \code{fastapath} or \code{bsgenome} is required to compute the frequency in
#'   sequences of each codon, used as normalization factors, even if \code{data}
#'   includes one or more columns among \emph{p_site_codon}, \emph{a_site_codon}
#'   and \emph{e_site_codon}. Default is NULL.
#' @param fasta_genome Logical value whether the FASTA file specified by
#'   \code{fastapath} contains nucleotide sequences of a genome assembly. If
#'   TRUE (the default), an annotation object is required (see \code{gtfpath}
#'   and \code{txdb}). FALSE implies nucleotide sequences of all transcripts are
#'   provided instead.
#' @param bsgenome Character string specifying the BSgenome data package with
#'   the genome sequences to be loaded. If not already present in the system, it
#'   is automatically installed through the biocLite.R script (check the list of
#'   available BSgenome data packages by running the
#'   \code{\link[BSgenome]{available.genomes}} function of the BSgenome
#'   package). This parameter must be coupled with an annotation object (see
#'   \code{gtfpath} and \code{txdb}). Please make sure the sequences included in
#'   the specified BSgenome data pakage are in agreement with the sequences used
#'   in the alignment step. Note: either \code{fastapath} or \code{bsgenome} is
#'   required to compute the frequency in sequences of each codon, used as
#'   normalization factors, even if \code{data} includes one or more columns
#'   among \emph{p_site_codon}, \emph{a_site_codon} and \emph{e_site_codon}.
#'   Default is NULL.
#' @param gtfpath Character string specifying the location of a GTF file,
#'   including its path, name and extension. Please make sure the GTF file and
#'   the sequences specified by \code{fastapath} or \code{bsgenome} derive from
#'   the same release. Note that either \code{gtfpath} or \code{txdb} is
#'   required if and only if nucleotide sequences of a genome assembly are
#'   provided (see \code{fastapath} or \code{bsgenome}). Default is NULL.
#' @param txdb Character string specifying the TxDb annotation package to be
#'   loaded. If not already present in the system, it is automatically installed
#'   through the biocLite.R script (check
#'   \href{http://bioconductor.org/packages/release/BiocViews.html#___TxDb}{here}
#'    the list of available TxDb annotation packages). Please make sure the TxDb
#'   annotation package and the sequences specified by \code{fastapath} or
#'   \code{bsgenome} derive from the same release. Note that either
#'   \code{gtfpath} or \code{txdb} is required if and only if nucleotide
#'   sequences of a genome assembly are provided (see \code{fastapath} or
#'   \code{bsgenome}). Default is NULL.
#' @param dataSource Optional character string describing the origin of the GTF
#'   data file. This parameter is considered only if \code{gtfpath} is
#'   specified. For more information about this parameter please refer to the
#'   description of \emph{dataSource} of the
#'   \code{\link[GenomicFeatures]{makeTxDbFromGFF}} function included in the
#'   \code{GenomicFeatures} package.
#' @param organism Optional character string reporting the genus and species of
#'   the organism of the GTF data file. This parameter is considered only if
#'   \code{gtfpath} is specified. For more information about this parameter
#'   please refer to the description of \emph{organism} of the
#'   \code{\link[GenomicFeatures]{makeTxDbFromGFF}} function included in the
#'   \code{GenomicFeatures} package.
#' @param transcripts Character string vector listing the name of transcripts to
#'   be included in the analysis. Default is NULL i.e. all transcripts are used.
#'   Please note: transcripts without annotated CDS and transcripts whose coding
#'   sequence length is not divisible by 3 are automatically discarded.
#' @param frequency_normalization Logical value whether to normalize the 64
#'   codon usage indexes for the corresponding codon frequencies in coding
#'   sequences. Default is TRUE.
#' @param codon_values Data table containing 64 triplet-specific values. If
#'   specified, the provided values are compared with the empirical codon usage
#'   indexes computed for the sample of interest. The data table must contain
#'   the DNA or RNA nucleotide sequence of the 64 codons and corresponding
#'   values arranged in two columns named \emph{codon} and \emph{value},
#'   respectively. Default is NULL.
#' @param label_scatter Logical value whether to label the dots in the scatter
#'   plot. Each dot is labeled using either the nucleotide sequence of the codon
#'   or the corresponding amino acid symbol (see \code{label_aminoacid}). This
#'   parameter is considered only if two sample names are specified in
#'   \code{sample} or \code{codon_values} is provided. Default is FALSE.
#' @param label_number Integer value in [1,64] specifying how many dots in the
#'   scatter plot should be labeled. Dots farthest from the confident interval
#'   of the regression line are automatically identified and labeled. Default is
#'   64 i.e. all dots are labeled. This parameter is considered only if
#'   \code{label_scatter} is TRUE.
#' @param label_aminoacid Logical value whether to use amino acid symbols to
#'   label the dots of the scatter. Default is FALSE i.e. codon nucleotide
#'   sequences are used instead. This parameter is considered only if
#'   \code{label_scatter} is TRUE.
#' @details \strong{riboWaltz} only works for read alignments based on
#'   transcript coordinates. This choice is due to the main purpose of RiboSeq
#'   assays to study translational events through the isolation and sequencing
#'   of ribosome protected fragments. Most reads from RiboSeq are supposed to
#'   map on mRNAs and not on introns and intergenic regions. Nevertheless, BAM
#'   based on transcript coordinates can be generated in two ways: i) aligning
#'   directly against transcript sequences; ii) aligning against standard
#'   chromosome sequences, requiring the outputs to be translated in transcript
#'   coordinates. The first option can be easily handled by many aligners (e.g.
#'   Bowtie), given a reference FASTA file where each sequence represents a
#'   transcript, from the beginning of the 5' UTR to the end of the 3' UTR. The
#'   second procedure is based on reference FASTA files where each sequence
#'   represents a chromosome, usually coupled with comprehensive gene annotation
#'   files (GTF or GFF). The STAR aligner, with its option --quantMode
#'   TranscriptomeSAM (see Chapter 6 of its
#'   \href{http://labshare.cshl.edu/shares/gingeraslab/www-data/dobin/STAR/STAR.posix/doc/STARmanual.pdf}{manual}),
#'    is an example of tool providing such a feature.
#' @return A list containing ggplot2 objects and a data table with the
#'   associated data ("dt"). If only one sample name is specified in
#'   \code{sample}, one ggplot2 object is returned (a bar plot named "plot"). If
#'   two sample names are specified in \code{sample}, three ggplot2 objects are
#'   returned (two bar plots named "plot_NameSample1" and
#'   "plot_NameSample2" and a scatter plot named "plot_comparison"). If
#'   \code{codon_values} is specified, two ggplot2 objects are returned (one bar
#'   plots named "plot" and a scatter plot named "plot_comparison"). Please
#'   note: before plotting, the 64 values are scaled to make them ranging
#'   between 0 and 1. If \code{frequency_normalization} is TRUE, the data table
#'   contains raw, normalized and scaled codon usage indexes, only raw and
#'   scaled data otherwise.
#' @examples
#' ## ## codon usage from transcriptome alignment
#' ## path_fasta <- "path/to/transcriptome/FASTA/file"
#' ## codon_usage_psite(data = reads_psite_list, annotation = mm81cdna,
#' ##                     sample = "Samp1",
#' ##                     fastapath = path_fasta, fasta_genome = FALSE)
#' ##
#' ## ## codon usage from genome alignment
#' ## path_fasta <- "path/to/genome/FASTA/file"
#' ## codon_usage_psite(data = reads_psite_list, annotation = mm81cdna,
#' ##                     sample = "Samp1",
#' ##                     fastapath = path_fasta, fasta_genome = TRUE)
#' @import data.table
#' @import ggplot2
#' @export
codon_usage_psite <- function(data, annotation, sample, site = "psite",
                              fastapath = NULL, fasta_genome = TRUE,
                              bsgenome = NULL, gtfpath = NULL, txdb = NULL, 
                              dataSource = NA, organism = NA, transcripts = NULL,
                              frequency_normalization = TRUE, codon_values = NULL,
                              label_scatter = FALSE, label_number = 64,
                              label_aminoacid = FALSE) {
  
  if(site != "psite" & site != "asite" & site != "esite"){
    cat("\n")
    stop("parameter site must be either \"psite\", \"asite\" or \"esite\" \n\n")
  }
  
  if(length(sample) > 2){
    cat("\n")
    stop("max two samples can be specified \n\n")
  }
  
  if(length(codon_values) != 0){
    rownames(codon_values) <- as.character(codon_values$codon)
    if (nrow(codon_values) < 64 | length(unique(rownames(codon_values))) < 64){
      cat("\n")
      stop("number of different triplets in codon_values < 64. At least one codon is missing\n\n")
    } else {
      if (length(unique(rownames(codon_values))) > 64){
        cat("\n")
        stop("number of different triplets in codon_values > 64. Too many codons\n\n")
      }
    }
  }
  
  if(length(sample) == 2 & length(codon_values) != 0){
    cat("\n")
    warning("codon_values will be ingnored: two samples are specified\n")
    codon_values = NULL
  }
  
  if((sum(lapply(data[sample], function(x) (gsub("site", "_site_codon", site) %in% colnames(x))) == TRUE) < length(sample)) |
     ((sum(lapply(data[sample], function(x) (gsub("site", "_site_codon", site) %in% colnames(x))) == TRUE) == length(sample)) &
      (frequency_normalization == TRUE | frequency_normalization == T))){
    
    if(length(fastapath) == 0 & length(bsgenome) == 0){
      cat("\n")
      if(sum(lapply(data[sample], function(x) (gsub("site", "_site_codon", site) %in% colnames(x))) == TRUE) < length(sample)){
        stop(paste0("unable to add ", gsub("site", "_site_codon", site), " column. Either fastpath or bsgenome must be specified\n\n"))
      } else {
        warning("no nucleotide sequences provided. Either fastpath or bsgenome must be specified. Non-normalized codon usage indexes will be returned\n")
        frequency_normalization = FALSE
      }
    }
    
    if(((length(fastapath) != 0 & (fasta_genome == TRUE | fasta_genome == T)) |
        length(bsgenome) != 0) &
       length(gtfpath) == 0 & length(txdb) == 0){
      cat("\n")
      stop("genome annotation file not specified. Either gtfpath or txdb object must be specified\n\n")
    }
    
    if(length(fastapath) != 0 & length(bsgenome) != 0){
      cat("\n")
      warning("both fastapath and bsgenome are specified. Only fastapath will be considered\n")
      bsgenome = NULL
    }
    
    if(length(gtfpath) != 0 & length(txdb) != 0){
      cat("\n")
      warning("both gtfpath and txdb are specified. Only gtfpath will be considered\n")
      txdb = NULL
    }
    
    if((length(gtfpath) != 0 | length(txdb) != 0) &
       ((length(fastapath) == 0 & length(bsgenome) == 0) |
        (length(fastapath) != 0 & (fasta_genome == FALSE | fasta_genome == F)))){
      cat("\n")
      stop("genome annotation file specified but no sequences from genome assembly provided\n\n")
    }
    
    if(length(gtfpath) != 0 | length(txdb) != 0){
      if(length(gtfpath) != 0){
        path_to_gtf <- gtfpath
        txdbanno <- GenomicFeatures::makeTxDbFromGFF(file = path_to_gtf, format = "gtf", dataSource = dataSource, organism = organism)
      } else {
        if(txdb %in% rownames(installed.packages())){
          library(txdb, character.only = TRUE)
        } else {
          source("https://bioconductor.org/biocLite.R")
          biocLite(txdb, suppressUpdates = TRUE)
          library(txdb, character.only = TRUE)
        }
        txdbanno <- get(txdb)
      }
    }
    
    if(length(fastapath) != 0 | length(bsgenome) != 0){
      if(length(fastapath) != 0) {
        if(fasta_genome == TRUE | fasta_genome == T){
          temp_sequences <- Biostrings::readDNAStringSet(fastapath, format = "fasta", use.names = TRUE)
          names(temp_sequences) <- tstrsplit(names(temp_sequences), " ", fixed = TRUE, keep = 1)[[1]]
          names(temp_sequences) <- tstrsplit(names(temp_sequences), "|", fixed = TRUE, keep = 1)[[1]]
          exon <- suppressWarnings(GenomicFeatures::exonsBy(txdbanno, by = "tx", use.names = TRUE))
          exon <- as.data.table(exon[unique(names(exon))])
          sub_exon_plus <- exon[as.character(seqnames) %in% names(temp_sequences) & strand == "+"]
          sub_exon_minus <- exon[as.character(seqnames) %in% names(temp_sequences) & strand == "-"
                                 ][, new_end := Biostrings::width(temp_sequences[as.character(seqnames)]) - start + 1
                                   ][, new_start := Biostrings::width(temp_sequences[as.character(seqnames)]) - end + 1]
          
          seq_dt_plus <- sub_exon_plus[, nt_seq := "emp"
                                       ][, nt_seq := as.character(Biostrings::subseq(temp_sequences[as.character(seqnames)],
                                                                                     start = start,
                                                                                     end = end))
                                         ][, list(seq = paste(nt_seq, collapse = "")), by = group_name]
          
          revcompl_temp_sequences <- Biostrings::reverseComplement(temp_sequences)
          seq_dt_minus <- sub_exon_minus[, nt_seq := "emp"
                                         ][, nt_seq := as.character(Biostrings::subseq(revcompl_temp_sequences[as.character(seqnames)],
                                                                                       start = new_start,
                                                                                       end = new_end))
                                           ][, list(seq = paste(nt_seq, collapse = "")), by = group_name]
          
          sequences <- Biostrings::DNAStringSet(c(seq_dt_plus$seq, seq_dt_minus$seq))
          names(sequences) <- c(unique(sub_exon_plus$group_name), unique(sub_exon_minus$group_name))
        } else {
          sequences <- Biostrings::readDNAStringSet(fastapath, format = "fasta", use.names = TRUE)
          names(sequences) <- tstrsplit(names(sequences), " ", fixed = TRUE, keep = 1)[[1]]
          names(sequences) <- tstrsplit(names(sequences), "|", fixed = TRUE, keep = 1)[[1]]
        }
      } else {
        if(bsgenome %in% installed.genomes()){
          library(bsgenome, character.only = TRUE)
        } else {
          source("http://www.bioconductor.org/biocLite.R")
          biocLite(bsgenome, suppressUpdates = TRUE)
          library(bsgenome, character.only = TRUE)
        }
        sequences <- GenomicFeatures::extractTranscriptSeqs(get(bsgenome), txdbanno, use.names = T)
      }
    }
    
    if(setequal(names(sequences), as.character(annotation$transcript)) == FALSE){
      if(length(setdiff(names(sequences), as.character(annotation$transcript))) > 0){
        cat("\n")
        warning("more sequences (FASTA) than transcripts (annotation)\n")
      } else {
        cat("\n")
        warning("more transcripts (annotation) than sequences (FASTA)\n")
      }
    }
    
    for(samp in sample){
      if(!(gsub("site", "_site_codon", site) %in% colnames(data[[samp]]))){
        if(site == "psite" & !("p_site_codon" %in% colnames(data[[samp]]))){
          data[[samp]] <- data[[samp]][, p_site_codon := as.character(Biostrings::subseq(sequences[as.character(data[[samp]]$transcript)],
                                                                                         start = data[[samp]]$psite,
                                                                                         end = data[[samp]]$psite + 2))]
        }
        if(site == "asite" & !("a_site_codon" %in% colnames(data[[samp]]))){
          data[[samp]] <- data[[samp]][, a_site_codon := as.character(Biostrings::subseq(sequences[as.character(data[[samp]]$transcript)],
                                                                                         start = data[[samp]]$psite + 3,
                                                                                         end = data[[samp]]$psite + 5))]
        }
        if(site == "esite" & !("e_site_codon" %in% colnames(data[[samp]]))){
          data[[samp]] <- data[[samp]][, e_site_codon := as.character(Biostrings::subseq(sequences[as.character(data[[samp]]$transcript)],
                                                                                         start = data[[samp]]$psite - 3,
                                                                                         end = data[[samp]]$psite - 1))]
        }
      }
    }
    
  }
  
  l_transcripts <- as.character(annotation[l_cds > 0 & l_cds %% 3 == 0, transcript])
  
  cod_aa <- data.table(codon=c("GCC", "GCG", "GCU", "GCA", "AGA", "CGG", "AGG", "CGA", "CGC", "CGU", "AAC", "AAU", "GAC", "GAU", "UGC", "UGU", "CAA", "CAG", "GAG", "GAA", "GGC", "GGU", "GGA", "GGG", "CAC", "CAU", "AUA", "AUC", "AUU", "CUG", "CUA", "UUA", "CUU", "UUG", "CUC", "AAA", "AAG", "AUG", "UUC", "UUU", "CCG", "CCC", "CCU", "CCA", "AGC", "UCG", "UCU", "UCA", "UCC", "AGU", "UAG", "UAA", "UGA", "ACA", "ACC", "ACG", "ACU", "UGG", "UAU", "UAC", "GUA", "GUG", "GUU", "GUC"),
                       aa=c("A", "A", "A", "A", "R", "R", "R", "R", "R", "R", "N", "N", "D", "D", "C", "C", "Q", "Q", "E", "E", "G", "G", "G", "G", "H", "H", "I", "I", "I", "L", "L", "L", "L", "L", "L", "K", "K", "M", "F", "F", "P", "P", "P", "P", "S", "S", "S", "S", "S", "S", "*", "*", "*", "T", "T", "T", "T", "W", "Y", "Y", "V", "V", "V", "V"))
  cod_lev <- gsub("U", "T", cod_aa$codon)
  
  if (length(transcripts) == 0) {
    c_transcript <- l_transcripts
  } else {
    c_transcript <- intersect(l_transcripts, transcripts)
  }
  
  if(frequency_normalization == TRUE | frequency_normalization == T){
    sub_sequences <- sequences[intersect(c_transcript, names(sequences))]
    cds_biost <- Biostrings::subseq(sub_sequences,
                                    start = annotation[transcript %in% names(sub_sequences), l_utr5] + 1,
                                    end = annotation[transcript %in% names(sub_sequences), l_utr5] +
                                      annotation[transcript %in% names(sub_sequences), l_cds])
    
    seq_freq <- (Biostrings::trinucleotideFrequency(cds_biost, step = 3,
                                                    as.prob = TRUE,
                                                    with.labels = TRUE,
                                                    simplify.as = "collapsed")) * 1000
    names(seq_freq) <- gsub("T", "U", names(seq_freq))
  }
  
  if(exists("norm_table")){
    rm(norm_table)
  }
  for(samp in sample) {
    dt <- data[[samp]][as.character(transcript) %in% c_transcript & psite_from_start %% 3 == 0]
    
    if(site == "psite"){
      dt <- dt[psite_region == "cds"]
      temp_table <- data.table(table(factor(dt$p_site_codon, levels = cod_lev)))
    } else {
      if(site == "asite"){
        dt <- dt[psite_from_start >= -3 & psite_from_stop <= -3]
        temp_table <- data.table(table(factor(dt$a_site_codon, levels = cod_lev)))
      } else {
        dt <- dt[psite_from_start >= 3 & psite_from_stop <= 3]
        temp_table <- data.table(table(factor(dt$e_site_codon, levels = cod_lev)))
      }
    }
    colnames(temp_table) <- c("codon", "raw_value")
    
    temp_table <- temp_table[, codon := gsub("T", "U", codon)]
    
    if(frequency_normalization == TRUE | frequency_normalization == T){
      temp_table <- temp_table[, normalized_value := raw_value / seq_freq[codon]]
      colplot <- "normalized_value"
    } else {
      colplot <- "raw_value"
    }
    
    temp_table <- temp_table[, plot_value := get(colplot) - min(get(colplot))
                             ][, plot_value := plot_value / max(plot_value)]
    
    if(!exists("norm_table")){
      norm_table <- temp_table
    } else {
      if(frequency_normalization == TRUE | frequency_normalization == T){
        norm_table[, "comp_normalized_value" := temp_table$normalized_value]
      }
      norm_table[, c("comp_raw_value", "comp_plot_value") := list(temp_table$raw_value, temp_table$plot_value)]
    }
  }
  
  norm_table <- norm_table[, class := "cds"
                           ][codon == "AUG", class := "start"
                             ][codon %in% c("UAA", "UGA", "UAG"), class := "stop"
                               ][, class := factor(class, levels = c("start", "cds", "stop"),
                                                   labels=c("Start codon", "cds", "Stop codon"))
                                 ][cod_aa, on = "codon"]
  
  if(length(sample) == 1){
    output <- list()
    if(frequency_normalization == TRUE | frequency_normalization == T){
      output[["dt"]] <- norm_table[, c("codon",  "aa", "class", "raw_value", "normalized_value", "plot_value")]
    } else {
      output[["dt"]] <- norm_table[, c("codon",  "aa", "class", "raw_value", "plot_value")]
    }
    
    if(length(codon_values) != 0){
      codon_values <- codon_values[, list(codon, value)]
      setnames(codon_values, old = c("codon", "value"), new =  c("codon", "comp_values"))
      codon_values[, codon := gsub("T", "U", codon)]
      norm_table <- norm_table[codon_values, on = "codon"
                               ][, comp_plot_value := comp_values - min(comp_values)
                                 ][, comp_plot_value := comp_plot_value / max(comp_plot_value)]
      
      output[["dt"]] <- cbind(output[["dt"]], norm_table[, c("comp_values", "comp_plot_value")])
      setnames(output[["dt"]], old = c("comp_values", "comp_plot_value"), new = c("user_value", "plot_user_value"))
      nc <- ncol(output[["dt"]])
      output[["dt"]] <- output[["dt"]][, c(1:(nc - 3), nc -1 , nc - 2, nc), with = FALSE]
      
      plotnamex <- "Codon usage index"
      plotnamey <- "User value"
    }
  } else {
    output <- list()
    if(frequency_normalization == TRUE | frequency_normalization == T){
      output[["dt"]] <- norm_table[, c("codon", "aa", "class", "raw_value", "comp_raw_value",
                                       "normalized_value", "comp_normalized_value", "plot_value", "comp_plot_value")]
      setnames(output[["dt"]],
               old = c("normalized_value", "comp_normalized_value"),
               new = paste0("normalized_value_", c(sample[1], sample[2])))
    } else {
      output[["dt"]] <- norm_table[, c("codon", "aa", "class", "raw_value", "comp_raw_value",
                                       "plot_value", "comp_plot_value")]
    }
    
    setnames(output[["dt"]],
             old = c("raw_value", "comp_raw_value", "plot_value", "comp_plot_value"),
             new = CJ(c("raw_value", "plot_value"), c(sample[1], sample[2]))[, paste(V1, V2, sep = "_")])
    
    plotnamex <- sample[1]
    plotnamey <- sample[2]
  }
  
  bs <- 30
  plot_list <- list()
  i = 1
  for(samp in sample){
    plot_col <- ifelse(i == 1, "plot_value", "comp_plot_value")
    i <- i + 1
    
    norm_table <- norm_table[order(get(plot_col))
                             ][, codon := factor(codon, levels = codon)]
    
    colour_codon <- ifelse(norm_table$codon == "AUG", "#104ec1",
                           ifelse(norm_table$codon %in% c("UAA", "UGA", "UAG"),
                                  "darkred", "gray40"))
    
    bp <- ggplot(norm_table, aes_string(x = "codon", y = plot_col, fill = "class")) +
      geom_bar(stat = "identity", alpha = 0.9) +
      scale_fill_manual(name = "", breaks=c("Start codon","Stop codon"), values = c("#104ec1", "gray60", "darkred")) +
      theme_bw(base_size = bs) +
      theme(legend.position = "top", legend.margin=margin(0,0,0,0), legend.box.margin=margin(5,0,-15,0)) +
      theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
      scale_x_discrete("Codon") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5, size = bs * 0.5)) +
      geom_text(aes(label = aa), vjust = -0.5, size = bs/6) +
      theme(axis.text.x = element_text(colour = colour_codon)) +
      scale_y_continuous("Usage index", limits = c(0,1.075), breaks = c(0,0.5,1))
    plot_list[[paste0("plot_", samp)]] <- bp
  }
  
  if(length(sample) == 1){
    output[["plot"]] <- plot_list[[paste0("plot_", sample)]]
  }
  
  if(length(codon_values) != 0 | length(sample) == 2){
    
    norm_table <- norm_table[order(plot_value)
                             ][, codon := factor(codon, levels = codon)]

    correlation <- round(cor(norm_table$plot_value, norm_table$comp_plot_value), 3)
    
    bs <- 25
    pcomp <- ggplot(norm_table,aes(x = plot_value, y = comp_plot_value, colour = class)) +
      geom_smooth(method = "lm", se=T, color="gray80", fill="gray80", linetype = 1, formula = y ~ x, level = 0.99,  fullrange = TRUE) +
      geom_point(alpha = 0.9, size = bs * 0.14) +
      scale_colour_manual(name = "", breaks = c("Start codon","Stop codon"), values = c("#104ec1", "gray40", "darkred")) +
      theme_bw(base_size = bs) +
      theme(legend.position = "top", legend.margin=margin(0,0,0,0), legend.box.margin=margin(5,0,-15,0)) +
      theme(legend.text = element_text(margin = margin(l = -12, unit = "pt"))) +
      scale_x_continuous(name = plotnamex, limits = c(-0.3,1.3), breaks = c(0,0.25,0.5,0.75,1), expand = c(0,0)) +
      scale_y_continuous(name = plotnamey, limits = c(-0.3,1.3), breaks = c(0,0.25,0.5,0.75,1), expand = c(0,0)) +
      coord_cartesian(xlim = c(-0.05,1.05), ylim = c(-0.05,1.05)) +
      annotate("text", x = 1, y = 0, label = paste0("R=",correlation), vjust = -0.2, size = bs * 0.2, hjust = 1, color = "black")
    
    if (label_scatter == T || label_scatter == TRUE) {
      if (label_number != 64){
        fit <- lm(norm_table$comp_plot_value ~ norm_table$plot_value)
        outlier_tab <- data.table(predict.lm(fit, interval = "confidence", level = 0.95)
        )[, codon := norm_table$codon
          ][, comp_plot_value := norm_table$comp_plot_value
            ][comp_plot_value < lwr | comp_plot_value > upr]
        outlier_cod <- as.character(outlier_tab[, max_dist := min(abs(comp_plot_value - lwr), abs(comp_plot_value - upr)),
                                                by = 1:nrow(outlier_tab)
                                                ][order(-max_dist)
                                                  ][1:label_number, codon])
        lab_table <- norm_table[, codon_out := ""
                                 ][as.character(codon) %in% outlier_cod, codon_out := codon
                                   ][, aa_out := ""
                                     ][codon_out != "", aa_out := aa]
        
        if (label_aminoacid == T || label_aminoacid == TRUE) {
          label_col <- "aa_out"
        } else {
          label_col <- "codon_out"
        }
      } else {
        if (label_aminoacid == T || label_aminoacid == TRUE) {
          label_col <- "aa"
        } else {
          label_col <- "codon"
        }
      }
      
      pcomp <- pcomp +
        ggrepel::geom_text_repel(data = lab_table, 
                                 aes_string("plot_value", "comp_plot_value", label = as.character(label_col)), show.legend = F)
    }
    
    if(length(sample) == 2){
      output[[paste0("plot_", sample[1])]] <- plot_list[[paste0("plot_", sample[1])]]
      output[[paste0("plot_", sample[2])]] <- plot_list[[paste0("plot_", sample[2])]]
    }
    
    output[["plot_comparison"]] <- pcomp
  }
  return(output)
}
